File: go/test/validate_format_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/eu-vat-validation-sdk/go"
	"github.com/voxgig-sdk/eu-vat-validation-sdk/go/core"

	vs "github.com/voxgig-sdk/eu-vat-validation-sdk/go/utility/struct"
)

func TestValidateFormatEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ValidateFormat(nil)
		if ent == nil {
			t.Fatal("expected non-nil ValidateFormatEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := validate_formatBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "validate_format." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set EUVATVALIDATION_TEST_VALIDATE_FORMAT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		validateFormatRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.validate_format", setup.data)))
		var validateFormatRef01Data map[string]any
		if len(validateFormatRef01DataRaw) > 0 {
			validateFormatRef01Data = core.ToMapAny(validateFormatRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = validateFormatRef01Data

		// LOAD
		validateFormatRef01Ent := client.ValidateFormat(nil)
		validateFormatRef01MatchDt0 := map[string]any{}
		validateFormatRef01DataDt0Loaded, err := validateFormatRef01Ent.Load(validateFormatRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if validateFormatRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func validate_formatBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "validate_format", "ValidateFormatTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read validate_format test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse validate_format test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"validate_format01", "validate_format02", "validate_format03", "country01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("EUVATVALIDATION_TEST_VALIDATE_FORMAT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"EUVATVALIDATION_TEST_VALIDATE_FORMAT_ENTID": idmap,
		"EUVATVALIDATION_TEST_LIVE":      "FALSE",
		"EUVATVALIDATION_TEST_EXPLAIN":   "FALSE",
		"EUVATVALIDATION_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["EUVATVALIDATION_TEST_VALIDATE_FORMAT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["EUVATVALIDATION_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["EUVATVALIDATION_APIKEY"],
			},
			extra,
		})
		client = sdk.NewEuVatValidationSDK(core.ToMapAny(mergedOpts))
	}

	live := env["EUVATVALIDATION_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["EUVATVALIDATION_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/validate_format_entity_test.rb
# ValidateFormat entity test

require "minitest/autorun"
require "json"
require_relative "../EuVatValidation_sdk"
require_relative "runner"

class ValidateFormatEntityTest < Minitest::Test
  def test_create_instance
    testsdk = EuVatValidationSDK.test(nil, nil)
    ent = testsdk.ValidateFormat(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = validate_format_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "validate_format." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set EU_VAT_VALIDATION_TEST_VALIDATE_FORMAT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    validate_format_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.validate_format")))
    validate_format_ref01_data = nil
    if validate_format_ref01_data_raw.length > 0
      validate_format_ref01_data = Helpers.to_map(validate_format_ref01_data_raw[0][1])
    end

    # LOAD
    validate_format_ref01_ent = client.ValidateFormat(nil)
    validate_format_ref01_match_dt0 = {}
    validate_format_ref01_data_dt0_loaded = validate_format_ref01_ent.load(validate_format_ref01_match_dt0, nil)
    assert !validate_format_ref01_data_dt0_loaded.nil?

  end
end

def validate_format_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "validate_format", "ValidateFormatTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = EuVatValidationSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["validate_format01", "validate_format02", "validate_format03", "country01"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["EU_VAT_VALIDATION_TEST_VALIDATE_FORMAT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "EU_VAT_VALIDATION_TEST_VALIDATE_FORMAT_ENTID" => idmap,
    "EU_VAT_VALIDATION_TEST_LIVE" => "FALSE",
    "EU_VAT_VALIDATION_TEST_EXPLAIN" => "FALSE",
    "EU_VAT_VALIDATION_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["EU_VAT_VALIDATION_TEST_VALIDATE_FORMAT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["EU_VAT_VALIDATION_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["EU_VAT_VALIDATION_APIKEY"],
      },
      extra || {},
    ])
    client = EuVatValidationSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["EU_VAT_VALIDATION_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["EU_VAT_VALIDATION_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
